File: app/src/main/java/de/eloc/eloc_control_panel/ng2/activities/DeleteAccountActivity.java
package de.eloc.eloc_control_panel.ng2.activities;

import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.hideKeyboard;
import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.open;
import static de.eloc.eloc_control_panel.ng3.activities.ActivityExtensionsKt.showModalAlert;

import android.os.Bundle;
import android.text.Editable;
import android.view.MenuItem;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.lifecycle.ViewModelProvider;

import de.eloc.eloc_control_panel.R;
import de.eloc.eloc_control_panel.databinding.ActivityDeleteAccountBinding;
import de.eloc.eloc_control_panel.ng3.activities.LoginActivity;
import de.eloc.eloc_control_panel.ng3.activities.ThemableActivity;
import de.eloc.eloc_control_panel.ng3.data.UserAccountViewModel;

public class DeleteAccountActivity extends ThemableActivity {

    private ActivityDeleteAccountBinding binding;
    private UserAccountViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityDeleteAccountBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        viewModel = new ViewModelProvider(this).get(UserAccountViewModel.class);

        setToolbar();
        setListeners();
        updateUI(false);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        int id = item.getItemId();
        if (id == android.R.id.home) {
            onBackPressed();
        }
        return true;
    }

    private void updateUI(boolean locked) {
        if (locked) {
            binding.passwordLayout.setEnabled(false);
            binding.submitButton.setEnabled(false);
            binding.progressHorizontal.setVisibility(View.VISIBLE);
        } else {
            binding.passwordLayout.setEnabled(true);
            binding.submitButton.setEnabled(true);
            binding.progressHorizontal.setVisibility(View.GONE);
        }
    }

    private void setToolbar() {
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(R.string.delete_account);
        }
    }

    private void setListeners() {
        binding.root.setOnClickListener(v -> hideKeyboard(this));
        binding.submitButton.setOnClickListener(v -> submit());
        binding.passwordTextInput.addTextChangedListener(new TextInputWatcher(binding.passwordLayout));
    }

    private void submit() {
        hideKeyboard(this);
        binding.passwordLayout.setError(null);

        String password = "";
        Editable editable = binding.passwordTextInput.getEditableText();
        if (editable != null) {
            password = editable.toString();
        }
        if (password.isEmpty()) {
            binding.passwordLayout.setError(getString(R.string.required));
            return;
        }

        updateUI(true);
        viewModel.verifyPassword(password, this::onPasswordVerificationCompleted);
    }

    private void onPasswordVerificationCompleted(String error) {
        if (error == null) {
            error = "";
        }
        error = error.trim();
        if (error.isEmpty()) {
            viewModel.deleteRemoteFiles(this::onRemoteFilesDeleted);
        } else {
            showModalAlert(
                    this,
                    getString(R.string.account),
                    error,
                    () -> updateUI(false)
            );
        }
    }

    private void onRemoteFilesDeleted(boolean success) {
        if (success) {
            viewModel.deleteProfile(this::onProfileDeleted);
        } else {
            showModalAlert(
                    this,
                    getString(R.string.oops),
                    getString(R.string.failed_to_remove_remote_files),
                    () -> updateUI(false)
            );
        }
    }

    private void onProfileDeleted(boolean success) {
        if (success) {
            viewModel.deleteAuthAccount(this::onAuthAccountDeleted);
        } else {
            showModalAlert(
                    this,
                    getString(R.string.oops),
                    getString(R.string.failed_to_clear_profile),
                    () -> updateUI(false)
            );
        }
    }

    private void onAuthAccountDeleted(boolean success) {
        if (success) {
            showModalAlert(
                    this,
                    getString(R.string.account),
                    "Your account has been deleted",
                    () -> {
                        viewModel.signOut();
                        open(this, LoginActivity.class, true);
                    }
            );
        } else {
            showModalAlert(
                    this,
                    getString(R.string.oops),
                    getString(R.string.failed_to_remove_account),
                    () -> updateUI(false)
            );
        }
    }
}

File: app/src/main/java/de/eloc/eloc_control_panel/ng3/activities/DeleteAccountActivity.kt
package de.eloc.eloc_control_panel.ng3.activities

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProvider
import de.eloc.eloc_control_panel.R
import de.eloc.eloc_control_panel.databinding.ActivityDeleteAccountBinding
import de.eloc.eloc_control_panel.ng2.activities.TextInputWatcher
import de.eloc.eloc_control_panel.ng3.data.UserAccountViewModel

class DeleteAccountActivity : ThemableActivity() {

    private lateinit var binding: ActivityDeleteAccountBinding
    private lateinit var viewModel: UserAccountViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityDeleteAccountBinding.inflate(layoutInflater)
        setContentView(binding.root)

        viewModel = ViewModelProvider(this)[UserAccountViewModel::class.java]

        setListeners()
        updateUI(false)
    }

    private fun updateUI(locked: Boolean) {
        if (locked) {
            binding.passwordLayout.isEnabled = false
            binding.submitButton.isEnabled = false
            binding.progressHorizontal.visibility = View.VISIBLE
        } else {
            binding.passwordLayout.isEnabled = true
            binding.submitButton.isEnabled = true
            binding.progressHorizontal.visibility = View.GONE
        }
    }

    private fun setListeners() {
        binding.root.setOnClickListener { hideKeyboard() }
        binding.submitButton.setOnClickListener { submit() }
        binding.passwordTextInput.addTextChangedListener(TextInputWatcher(binding.passwordLayout))
        binding.elocAppBar.setOnBackButtonClickedListener { goBack() }
    }

    private fun submit() {
        hideKeyboard()
        binding.passwordLayout.error = null

        val password = binding.passwordTextInput.editableText?.toString() ?: ""
        if (password.isEmpty()) {
            binding.passwordLayout.error = getString(R.string.required)
            return
        }

        updateUI(true)
        viewModel.verifyPassword(password, this::onPasswordVerificationCompleted)
    }

    private fun onPasswordVerificationCompleted(err: String) {
        val error = err.trim()
        if (error.isEmpty()) {
            viewModel.deleteRemoteFiles(this::onRemoteFilesDeleted)
        } else {
            showModalAlert(getString(R.string.account), error) { updateUI(false) }
        }
    }

    private fun onRemoteFilesDeleted(success: Boolean) {
        if (success) {
            viewModel.deleteProfile(this::onProfileDeleted)
        } else {
            showModalAlert(
                getString(R.string.oops),
                getString(R.string.failed_to_remove_remote_files)
            ) { updateUI(false) }
        }
    }

    private fun onProfileDeleted(success: Boolean) {
        if (success) {
            viewModel.deleteAuthAccount(this::onAuthAccountDeleted)
        } else {
            showModalAlert(
                getString(R.string.oops),
                getString(R.string.failed_to_clear_profile),
            ) { updateUI(false) }
        }
    }

    private fun onAuthAccountDeleted(success: Boolean) {
        if (success) {
            showModalAlert(
                getString(R.string.account),
                getString(R.string.your_account_has_been_deleted)
            ) {
                viewModel.signOut()
                open(LoginActivity::class.java, true)
            }
        } else {
            showModalAlert(
                getString(R.string.oops),
                getString(R.string.failed_to_remove_account),
            ) { updateUI(false) }

        }
    }
}